panther_hardware_interfaces/src/panther_system.cpp

```diff
@@ -548,6 +606,82 @@ bool PantherSystem::CheckIfSafetyStopActive()
          rear_data.GetRightRuntimeError().GetMessage().safety_stop_active;
 }
 
+bool PantherSystem::AreVelocityCommandsNearZero()
+{
+  for (const auto & cmd : hw_commands_velocities_) {
+    if (std::abs(cmd) > std::numeric_limits<double>::epsilon()) {
+      return false;
+    }
+  }
+  return true;
+}
+
+void PantherSystem::SetEStop()
+{
+  RCLCPP_INFO(logger_, "Setting E-stop");
+  bool gpio_controller_error = false;
+
+  try {
+    gpio_controller_->EStopTrigger();
+  } catch (const std::runtime_error & e) {
+    RCLCPP_ERROR_STREAM(logger_, "Error when trying to set E-stop stop using GPIO: " << e.what());
+    gpio_controller_error = true;
+  }
+
+  try {
+    {
+      std::lock_guard<std::mutex> lck_g(motor_controller_write_mtx_);
+      motors_controller_->TurnOnSafetyStop();
+    }
+  } catch (const std::runtime_error & e) {
+    RCLCPP_ERROR_STREAM(
+      logger_, "Error when trying to set safety stop using CAN command: " << e.what());
+    if (gpio_controller_error) {
+      RCLCPP_ERROR_STREAM(logger_, "Both attempts at setting E-stop failed");
+      throw std::runtime_error("Both attempts at setting E-stop failed");
+    }
+  }
+
+  e_stop_ = true;
+}
+
+void PantherSystem::ResetEStop()
+{
+  RCLCPP_INFO(logger_, "Resetting E-stop");
+
+  // On side of the motors controller safety stop is reset by sending 0.0 commands
+  if (!last_commands_zero_) {
+    RCLCPP_ERROR(
+      logger_,
+      "Can't reset E-stop - last velocity commands are different than zero. Make sure that your "
+      "controller sends zero commands before trying to reset E-stop.");
+    throw std::runtime_error(
+      "Can't reset E-stop - last velocity commands are different than zero. Make sure that your "
+      "controller sends zero commands before trying to reset E-stop.");
+  }
+
+  try {
+    gpio_controller_->EStopReset();
+  } catch (const std::runtime_error & e) {
+    RCLCPP_ERROR_STREAM(logger_, "Error when trying to reset E-stop using GPIO: " << e.what());
+    throw e;
+  }
+
+  roboteq_error_filter_->SetClearErrorsFlag();
+  e_stop_ = false;
+}
+
+bool PantherSystem::ReadEStop()
+{
+  if (panther_version_ >= 1.2 - std::numeric_limits<float>::epsilon()) {
+    // TODO: it has reversed logic
+    return !gpio_controller_->IsPinActive(panther_gpiod::GPIOPin::E_STOP_RESET);
```

Comment:
I thought about changing the `active_low` setting, but when I reconsidered, It makes more sense that EStop is true when EStop reset is false, so we can drop it.